ql/lib/codeql/kaleidoscope/Ast.qll

```diff
@@ -0,0 +1,410 @@
+
+  override Location getLocation() { result = node.getLocation() }
+
+  Expression getCondition() { toTreeSitter(result) = node.getCondition() }
+
+  Expression getThen() { toTreeSitter(result) = node.getThen() }
+
+  Expression getElse() { toTreeSitter(result) = node.getElse() }
+}
+
+class ForExpression extends Expression, TForExpression {
+  private Kaleidoscope::ForExpression node;
+
+  ForExpression() { this = TForExpression(node) }
+
+  override AstNode getAChild(string name) {
+    name = "getInitializer" and result = this.getInitializer()
+    or
+    name = "getCondition" and result = this.getCondition()
+    or
+    name = "getStep" and result = this.getStep()
+    or
+    name = "getBody" and result = this.getBody()
+  }
+
+  override string toString() { result = "for ... in ..." }
+
+  override string getAPrimaryQlClass() { result = "ForExpression" }
+
+  override Location getLocation() { result = node.getLocation() }
+
+  Initializer getInitializer() { toTreeSitter(result) = node.getInitializer() }
+
+  Expression getCondition() { toTreeSitter(result) = node.getCondition() }
+
+  Expression getStep() {
+    toTreeSitter(result) = node.getUpdate()
+    or
+    not exists(node.getUpdate()) and
+    result = TDefaultStep(any(Kaleidoscope::ReservedWord i | i.getParent() = node))
+  }
+
+  Expression getBody() { toTreeSitter(result) = node.getBody() }
+}
+
+class FunctionCallExpression extends Expression, TFunctionCallExpression {
+  private Kaleidoscope::FunctionCallExpression node;
+
+  FunctionCallExpression() { this = TFunctionCallExpression(node) }
+
+  override AstNode getAChild(string name) {
+    name = "getCallee" and result = this.getCallee()
+    or
+    name = "getArgument" and result = this.getArgument(_)
+  }
+
+  override string toString() { result = this.getCallee().getName() + "(...)" }
+
+  override string getAPrimaryQlClass() { result = "FunctionCallExpression" }
+
+  override Location getLocation() { result = node.getLocation() }
+
+  Identifier getCallee() { toTreeSitter(result) = node.getCallee() }
+
+  Expression getArgument(int index) { toTreeSitter(result) = node.getArgument(index) }
+}
+
+private class TNumber = TDefaultStep or TNumberLiteral;
+
+class Number extends Expression, TNumber {
+  override string getAPrimaryQlClass() { result = "NumberLiteral" }
+}
+
+private class NumberLiteral extends Number, TNumberLiteral {
+  private Kaleidoscope::Number node;
+
+  NumberLiteral() { this = TNumberLiteral(node) }
+
+  override string toString() { result = node.getValue() }
+
+  override Location getLocation() { result = node.getLocation() }
+}
+
+private class DefaultStep extends Number, TDefaultStep {
+  private Kaleidoscope::ReservedWord node;
+
+  DefaultStep() { this = TDefaultStep(node) }
+
+  override string toString() { result = "1.0" }
+
+  override Location getLocation() { result = node.getLocation() }
+}
+
+class ParenExpression extends Expression, TParenExpression {
```

Comment:
Did you consider using the synthesis/desugaring framework for desugaring away parentheses?

Reply:
That felt a bit overkill .